ibis/expr/analysis.py

```diff
@@ -1039,75 +1036,41 @@ def is_scalar_reduction(expr):
     return isinstance(expr, ir.Scalar) and is_reduction(expr)
 
 
-class _AnyToExistsTransform:
-    __slots__ = (
-        "expr",
-        "parent_table",
-        "query_roots",
-        "foreign_table",
-        "predicates",
-    )
-
-    def __init__(self, expr: ir.Expr, parent_table: ir.Table) -> None:
-        self.expr = expr
-        self.parent_table = parent_table
-        self.query_roots: AbstractSet[ops.TableNode] = frozenset(
-            self.parent_table.op().root_tables()
-        )
-        self.foreign_table: ir.Table | None = None
-        self.predicates: MutableSequence[ir.BooleanValue] = []
-
-    def get_result(self) -> ir.BooleanColumn:
-        self._visit(self.expr)
+_ANY_OP_MAPPING = {
+    ops.Any: ops.UnresolvedExistsSubquery,
+    ops.NotAny: ops.UnresolvedNotExistsSubquery,
+}
 
-        op = self.expr.op()
 
-        if isinstance(op, ops.Any):
-            op_class = ops.ExistsSubquery
-        else:
-            assert isinstance(op, ops.NotAny)
-            op_class = ops.NotExistsSubquery
+def find_predicates(expr, flatten=True):
+    def predicate(expr):
+        if isinstance(expr, ir.BooleanColumn):
+            if flatten and isinstance(expr.op(), ops.And):
+                return lin.proceed, None
+            else:
+                return lin.halt, expr
+        return lin.proceed, None
 
-        return op_class(self.foreign_table, self.predicates).to_expr()
+    return list(lin.traverse(predicate, expr))
 
-    def _visit(self, expr: ir.Expr) -> None:
-        node = expr.op()
 
-        for arg in node.flat_args():
-            if isinstance(arg, ir.Table):
-                self._visit_table(arg)
-            elif isinstance(arg, ir.BooleanColumn):
-                for sub_expr in flatten_predicate(arg):
-                    self.predicates.append(sub_expr)
-                    self._visit(sub_expr)
-            elif isinstance(arg, ir.Expr):
-                self._visit(arg)
-
-    def _find_blocking_table(self, expr: ir.Expr) -> ir.Expr:
-        node = expr.op()
-
-        if node.blocks():
-            return expr
+def _make_any(
+    expr,
+    any_op_class: type[ops.Any] | type[ops.NotAny],
+):
+    tables = list(find_immediate_parent_tables(expr))
+    predicates = find_predicates(expr, flatten=True)
 
-        return next(
-            (
-                result
-                for arg in node.flat_args()
-                if (
-                    isinstance(arg, ir.Expr)
-                    and (result := self._find_blocking_table(arg)) is not None
-                )
-            ),
-            None,
+    ntables = len(tables)
+    if (ntables := len(tables)) == 1:
```

Comment:
Use `analysis.has_multiple_bases()`?

Reply:
> Use `analysis.has_multiple_bases()`?

Doesn't that mean I would traverse the expr twice? I've already got the table count, so there's no point in doing that extra computation.